### PR TITLE
Enable dynamo traced test_forloop_goes_right_direction

### PR DESCRIPTION
**Pull request by mlazos**
Removed a bunch of skips, I also updated test_forloop_goes_right_direction to *not* use the closure when dynamo is tracing. The reason for this is that testing the disabled optimizer doesn't actually test anything.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123469
* #123619
* #123406
* #124309
* #123405
* #123404
* #123324
* #123323
* __->__ #123322
* #123498

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: test/test_optim.py [hunk 1]
```diff
@@ -112,7 +112,15 @@ def closure():
 
             initial_value = closure().item()
             for _ in range(20):
-                loss = optimizer.step(closure)
+                if torch._dynamo.is_compiling() and optim_cls != torch.optim.LBFGS:
```
Comment (janeyx99) [inline test/test_optim.py:115]: Use the optim_info.step_requires_closure flag instead. Can you add a separate test to test correctness of closures under torch.compile in test_compiled_optimizers.py (that it falls back and computes the same thing)?

**Review by janeyx99 — COMMENTED** [review]
Could you add more detail that the enablement is for dynamo testing

File: test/inductor/test_compiled_optimizers.py [hunk 2]
```diff
@@ -343,66 +342,75 @@ def test_correctness(self, device, dtype, optim_info):
...
-            opt_compiled = optim_cls(model_compiled.parameters(), **kwargs)
-            opt_eager = optim_cls(model_eager.parameters(), **kwargs)
-
-            if optim_cls is LBFGS:
-
-                @torch.compile()
-                def fn():
-                    def closure():
-                        loss = model_compiled(input).sum()
+        closure_options = (True,) if optim_info.step_requires_closure else (False, True)
```
Comment (janeyx99) [inline test/inductor/test_compiled_optimizers.py:345]: Better to use parametrization over the test similar to with_lrsched here: https://github.com/pytorch/pytorch/blob/main/test/test_optim.py#L82

It is easier to read and will make it separate tests while allowing you to reuse the code

File: test/inductor/test_compiled_optimizers.py [hunk 3]
```diff
@@ -343,66 +342,75 @@ def test_correctness(self, device, dtype, optim_info):
...
+
+                        opt_compiled.step(closure)
+
+                    def closure_eager():
+                        loss = model_eager(input).sum()
                         loss.backward()
-                        return loss
-
-                    opt_compiled.step(closure)
+                        if optim_cls is SparseAdam:
```
Comment (janeyx99) [inline test/inductor/test_compiled_optimizers.py:386]: Use the supports_only_sparse_grads flag

File: test/inductor/test_compiled_optimizers.py (hunk 2 above)
Comment (mlazos) [reply test/inductor/test_compiled_optimizers.py:345]: I would like to skip the parameterization for the requires closure optimizers though. Don't see a way to do that with parameterizations..

File: test/test_optim.py [hunk 4]
```diff
@@ -112,7 +112,15 @@ def closure():
...
-                loss = optimizer.step(closure)
+                if torch._dynamo.is_compiling() and not optim_info.step_requires_closure:
+                    # we disable torch.compile with the closure argument, so in order to actually
+                    # test compiling the optimizer, we separate out the closure here.
+                    # LBFGS requires the closure arg but torch.compile doesn't support LBFGS
+                    # so it's fine to run the disabled path.
+                    loss = closure()
+                    optimizer.step()
+                else:
+                    loss = optimizer.step(closure)
```
Comment (janeyx99) [inline test/test_optim.py:123]: It's fine to just flip this to
```suggestion
                if optim_info.step_requires_closure:
                    loss = optimizer.step(closure)
                else:
                    # we disable torch.compile with the closure argument, so in order to actually
                    # test compiling the optimizer, we separate out the closure here.
                    loss = closure()
                    optimizer.step()
                    
```

**Review by janeyx99 — APPROVED** [review]
Approving, but use my suggestion to minimize disparity between eager and dynamo tests,

**Pull request closed without merging**